types/src/v0/impls/stake_table.rs

```diff
@@ -0,0 +1,347 @@
+use super::{L1Client, NodeState, PubKey, SeqTypes};
+use contract_bindings::permissioned_stake_table::{NodeInfo, StakersUpdatedFilter};
+use ethers::{abi::Address, types::U256};
+use hotshot::types::SignatureKey;
+use hotshot_contract_adapter::stake_table::NodeInfoJf;
+use hotshot_types::{
+    stake_table::StakeTableEntry,
+    traits::{
+        election::Membership, node_implementation::NodeType, signature_key::StakeTableEntryType,
+    },
+    PeerConfig,
+};
+use std::{
+    cmp::max,
+    collections::{BTreeMap, HashSet},
+    num::NonZeroU64,
+    str::FromStr,
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Clone, Debug)]
+pub struct StaticCommittee<T: NodeType> {
+    /// The nodes eligible for leadership.
+    /// NOTE: This is currently a hack because the DA leader needs to be the quorum
+    /// leader but without voting rights.
+    eligible_leaders: Vec<<T::SignatureKey as SignatureKey>::StakeTableEntry>,
+
+    /// The nodes on the committee and their stake
+    stake_table: HashSet<<T::SignatureKey as SignatureKey>::StakeTableEntry>,
+
+    /// The nodes on the committee and their stake
+    da_stake_table: HashSet<<T::SignatureKey as SignatureKey>::StakeTableEntry>,
+
+    /// The nodes on the committee and their stake, indexed by public key
+    indexed_stake_table:
+        BTreeMap<T::SignatureKey, <T::SignatureKey as SignatureKey>::StakeTableEntry>,
+
+    /// The nodes on the committee and their stake, indexed by public key
+    indexed_da_stake_table:
+        BTreeMap<T::SignatureKey, <T::SignatureKey as SignatureKey>::StakeTableEntry>,
+
+    /// Number of blocks in an epoch
+    epoch_size: u64,
+
+    /// Address of StakeTable contract (proxy address)
+    contract_address: Option<Address>,
+
+    /// L1 provider
+    provider: L1Client,
+}
+
+impl<TYPES: NodeType> StaticCommittee<TYPES> {
+    /// Updates `Self.stake_table` with stake_table for
+    /// `Self.contract_address` at `l1_block_height`. This is intended
+    /// to be called before calling `self.stake()` so that
+    /// `Self.stake_table` only needs to be updated once in a given
+    /// life-cycle but may be read from many times.
+    async fn update_stake_table(&mut self, l1_block_height: u64) {
+        // TODO also deal w/ removed entries
+        let updates: Vec<StakersUpdatedFilter> = self
+            .provider
+            .get_stake_table::<TYPES>(l1_block_height, self.contract_address.unwrap())
+            .await;
+
+        let added: Vec<NodeInfoJf> = updates
+            .into_iter()
+            .flat_map(|e: StakersUpdatedFilter| e.added.into_iter().map(NodeInfoJf::from))
+            .collect();
+        for node in added {
+            if !node.da {
+                let entry: <TYPES::SignatureKey as SignatureKey>::StakeTableEntry =
+                    StakeTableEntry {
```

Comment:
Need to match up types here.

Reply:
https://github.com/EspressoSystems/espresso-sequencer/commit/ef90df412caf7a0757dfc7b6387a22471136f8ff